### PR TITLE
Add eject handling question support

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Ref: https://github.com/terraform-providers/terraform-provider-vcd/issues/170

After some time figure out that there is possibility to query and answer questions for VM then ejecting mounted media. vCD ask question do you really want eject mounted media when VM is runing. This functionality will be used to improve current existing terraform resource.

Improves API with new functionality:
* new functions to handle VM questions and answers
* eject task which handles media eject question and answers during task
* removed check for VM to be power off for ejecting
* new unitest

### CODE REVIEW

File: types/v56/types.go [hunk 1]
```diff
@@ -2274,3 +2274,35 @@ type MediaInsertOrEjectParams struct {
...
+// Parameters for VM question answer choice
+// Reference: vCloud API 27.0 - VmQuestionAnswerChoiceType
+// https://code.vmware.com/apis/287/vcloud#/doc/doc/types/VmQuestionAnswerChoiceType.html
+type VmQuestionAnswerChoiceType struct {
+	Id   int    `xml:"Id"`
+	Text string `xml:"Text,omitempty"`
+}
+
+// Parameters for VM question answer
+// Reference: vCloud API 27.0 - VmPendingQuestionType
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply types/v56/types.go]: Fixed

File: govcd/ejecttask.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+		// If task is not in a waiting status we're done, check if there's an error and return it.
+		if ejectTask.Task.Task.Status != "queued" && ejectTask.Task.Task.Status != "preRunning" && ejectTask.Task.Task.Status != "running" {
+			if ejectTask.Task.Task.Status == "error" {
+				return fmt.Errorf("task did not complete succesfully: %s", ejectTask.Task.Task.Error.Message)
+			}
+			return nil
+		}
+
+		question, err := ejectTask.vm.GetQuestions()
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/ejecttask.go]: I verifyed and yes it singular. Fixed

File: govcd/vm.go [hunk 3]
```diff
@@ -693,3 +690,71 @@ func (vm *VM) insertOrEjectMedia(mediaParams *types.MediaInsertOrEjectParams, li
 
 	return *task, nil
 }
+
+// Use the get existing VM questions for operations which need additional response
+// Reference:
+// https://code.vmware.com/apis/287/vcloud#/doc/doc/operations/GET-VmPendingQuestion.html
+func (vm *VM) GetQuestions() (types.VmPendingQuestion, error) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: Fixed

File: govcd/vm.go [hunk 4]
```diff
@@ -693,3 +690,71 @@ func (vm *VM) insertOrEjectMedia(mediaParams *types.MediaInsertOrEjectParams, li
...
+
+	apiEndpoint, _ := url.ParseRequestURI(vm.VM.HREF)
+	apiEndpoint.Path += "/question"
+
+	req := vm.client.NewRequest(map[string]string{}, "GET", *apiEndpoint, nil)
+
+	resp, err := checkResp(vm.client.Http.Do(req))
+
+	// vCD security feature - on no questions return 403 access error
+	if err != nil && strings.Contains(err.Error(), "API Error: 403: The access to the resource") {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: Thank tou for suggestions, I changed according with just right order.

File: govcd/vm.go [hunk 5]
```diff
@@ -693,3 +690,71 @@ func (vm *VM) insertOrEjectMedia(mediaParams *types.MediaInsertOrEjectParams, li
...
+	return *question, nil
+
+}
+
+// Use the provide answer to existing VM questions for operations which need additional response
+// Reference:
+// https://code.vmware.com/apis/287/vcloud#/doc/doc/operations/POST-AnswerVmPendingQuestion.html
+func (vm *VM) AnswerQuestion(questionId string, choiceId int) error {
+
+	//validate input
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: Added missing validation

File: govcd/vm.go [hunk 6]
```diff
@@ -583,10 +584,10 @@ func (vm *VM) HandleInsertMedia(org *Org, catalogName, mediaName string) (Task,
 }
 
 // Helper function which finds media and calls EjectMedia
-func (vm *VM) HandleEjectMedia(org *Org, catalogName, mediaName string) (Task, error) {
+func (vm *VM) HandleEjectMedia(org *Org, catalogName, mediaName string) (EjectTask, error) {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/vm.go]: Yes it's true, but anyways I think it's safe as this functionality is brand new and work on right in limited cases.

File: govcd/ejecttask.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		if err != nil {
+			return fmt.Errorf("task did not complete succesfully: %s, quering question for VM failed: %s", ejectTask.Task.Task.Description, err.Error())
+		}
+
+		if question.QuestionId != "" && strings.Contains(question.Question, "Disconnect anyway and override the lock?") {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/ejecttask.go]: We are not confident at all, but we don't have any other possible hooking possibilities. Somehow we have differentiate questions and in xml there is only one thing - text.

File: govcd/ejecttask.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	return &EjectTask{
+		task,
+		vm,
+	}
+}
+
+// Checks the status of the task every 3 seconds and returns when the
+// eject task is either completed or failed
+func (ejectTask *EjectTask) WaitTaskCompletion(isAnswerYes bool) error {
+	return ejectTask.WaitInspectTaskCompletion(isAnswerYes, 3*time.Second)
```
Comment (lvirbalas) [inline govcd/ejecttask.go]: Please expose this retry interval variable at the top of the file, so everyone notices that there's some waiting game involved.
Also, if the call is not too expensive let's make it check every single second as opposed to three.

File: govcd/ejecttask.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+			}
+			return nil
+		}
+
+		question, err := ejectTask.vm.GetQuestion()
+		if err != nil {
+			return fmt.Errorf("task did not complete succesfully: %s, quering question for VM failed: %s", ejectTask.Task.Task.Description, err.Error())
+		}
+
+		if question.QuestionId != "" && strings.Contains(question.Question, "Disconnect anyway and override the lock?") {
```
Comment (lvirbalas) [inline govcd/ejecttask.go]: As in above comment, please expose this string at the top of the file, so it's easy to notice the hardcoded thing.

File: govcd/ejecttask.go (hunk 8 above)
Comment (vbauzysvmware) [reply govcd/ejecttask.go]: Changed

File: govcd/ejecttask.go (hunk 9 above)
Comment (vbauzysvmware) [reply govcd/ejecttask.go]: Changed

**Comment by vbauzysvmware** [comment]
Added CHANGELOG file

**Pull request merged**